src/domain/mongoose_domain_api.erl

```diff
@@ -98,8 +107,10 @@ delete_domain(Domain, HostType) ->
 
 %% This is ran only in the context of `do_delete_domain',
 %% so it can already skip some checks
--spec do_delete_domain_in_progress(domain(), host_type()) -> delete_domain_return().
-do_delete_domain_in_progress(Domain, HostType) ->
+-spec do_delete_domain_in_progress(domain(), host_type(), sync | async) -> delete_domain_return().
+do_delete_domain_in_progress(Domain, HostType, async) ->
```

Comment:
Shouldn't async operation block sync operation? Currently, we allow to do both operations at the same time (or am I missing something?) and it's hard to say what output would we receive.

Generally, I wonder if it wouldn't be better to replace the old endpoint with a new one. After merging this PR, we will have two endpoints for the same operation where one (sync) isn't working correctly in all cases.

Reply:
We allow both, but they happen in different processes. A sync operation would run on the cowboy process that was spawned for this request, while in an async operation the cowboy process would only synchronously mark the domain as `deleting` and then submit a request to the cleaner to do the cleaning offline, so that the cowboy process can finish its work and return. If you were to do both operations concurrently, the same caveats apply than doing two deletes in the old sync way concurrently, after all they also spawn two parallel cowboy processes to handle the request.

The sync operation is theoretically correct, the problem is that most often HTTP clients will have a timeout and kill the request before the sync operation can finish. But in that case the operation can be retried, that was added in a previous PR. Or the client can submit a much longer timeout than usual.

About whether to remove the sync operation, perhaps, dunno, I'm kinda in favour of leaving it because a client might want to know promptly that the operation is done, but I don't really have an opinion here, don't know what's the best API practice for this matter.